Add the possibility to pass custom front- and backend event exception handlers to the rx.App

maximvlah opened the pull request:
### All Submissions:

- [x] Have you followed the guidelines stated in [CONTRIBUTING.md](https://github.com/reflex-dev/reflex/blob/main/CONTRIBUTING.md) file?
- [x] Have you checked to ensure there aren't any other open [Pull Requests](https://github.com/reflex-dev/reflex/pulls ) for the desired changed?

### Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### New Feature Submission:

- [x] Does your submission pass the tests? 
- [x] Have you linted your code locally prior to submission?

### Changes To Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?


```python
#Test App

import reflex as rx
import traceback

class TestAppState(rx.State):
    """State for the TestApp app."""

    def divide_by_number(self, number: int):
        """Divide by number and print the result.

        Args:
            number: number to divide by
        """
        print(1 / number)


def frontend_exception_handler(exception: Exception):
    """Frontend exception handler."""

    print(f"Frontend exception: {exception}")

def backend_exception_handler(exception: Exception):
    """Backend exception handler."""

    error = traceback.format_exc()

    print(f"Backend exception: {error}")


app = rx.App(
    state=rx.State, 
    frontend_exception_handler=frontend_exception_handler, 
    backend_exception_handler=backend_exception_handler
)

@app.add_page
def index():
    return rx.vstack(
        rx.button(
            "induce_frontend_error",
            on_click=rx.call_script("induce_frontend_error()"),
            id="induce-frontend-error-btn",
        ),
        rx.button(
            "induce_backend_error",
            on_click=lambda: TestAppState.divide_by_number(0),  # type: ignore
            id="induce-backend-error-btn",
        ),
    )
    
```

closes #3284

masenf commented on reflex/state.py:
```diff
@@ -1812,6 +1814,20 @@ class State(BaseState):
     # The hydrated bool.
     is_hydrated: bool = False
 
+    def handle_frontend_exception(self, stack: str) -> None:
```
this should be moved into its own substate like the other built in functionality defined below.

having it here means that any frontend exception will trigger a load of the entire State tree with all substates. if it's in its own substate, then only that substate will get loaded

masenf reviewed (commented):
thanks for the contribution. i'll test it out further this week

Lendemor reviewed (changes requested):
I've started testing it on my side.

No issue with the backend error handler, seems to work properly.

I've had issue triggering it for frontend error however, seems like not all errors are being caught?

The following code still produce a frontend error, but doesn't seems to send it to the backend.
ie `[Reflex Frontend Exception]` does not appear in that case.
```python
class MockCompo(rx.Component):
    tag = "UnknownTag"

@rx.page()
def index():
    return rx.foreach([1, 2, 3], MockCompo.create)
```

Also tested with 
```python
@rx.page()
def index():
    return rx.script(console.log(foo))
```
In this case it also doesn't triggers.


Also tested with 
```python
@rx.page()
def index():
    return rx.button("Test", on_click=rx.call_script("console.log(foo)"))
```
In this case it does trigger the frontend handler.

Lendemor commented:
Also I think the default handler should be tied to the `App`, not to the `Config`.

Config should stay as simple as possible.

maximvlah commented:
@Lendemor thanks for the feedback. I will continue working to handle aforementioned edge cases tomorrow. In the meantime, I have just moved the exception handlers from rx.Config to rx.App as you demanded.

masenf reviewed (commented):
This seems to work fine in dev mode (although a frontend error triggers the handler twice for some reason).

But in prod mode, when i hit an unhandled frontend error, it does not fire the event...

```python
import reflex as rx


class State(rx.State):
    broken: bool = False


def index() -> rx.Component:
    return rx.container(
        rx.vstack(
            rx.cond(
                State.broken,
                rx.text("Broken", size=rx.Var.create("x", _var_is_local=False, _var_is_string=False)),
            ),
            rx.text("Break the app? ", rx.switch(checked=State.broken, on_change=State.set_broken)),
            spacing="5",
            justify="center",
            min_height="85vh",
        ),
        rx.logo(),
    )


app = rx.App()
app.add_page(index)
```

Flipping the switch introduces a component with a rendering error.

In dev mode, flipping the switch prints the error to the terminal, browser console, and on screen.

In prod mode, flipping the switch redirects to a page that says "Application error: a client-side exception has occurred (see the browser console for more information)."

maximvlah commented on reflex/state.py:
```diff
@@ -1093,6 +1093,9 @@ def __setattr__(self, name: str, value: Any):
         # Set the var on the parent state.
         inherited_vars = {**self.inherited_vars, **self.inherited_backend_vars}
         if name in inherited_vars:
+            if self.get_name() == "frontend_event_exception_state":
+                return
```
Because when running Reflex with Redis, I get the error below. Without Redis, everything works fine. Do you know why? You can test it in the integration/test_exception_handlers.py

```bash

ERROR    asyncio:base_events.py:1821 Task exception was never retrieved
future: <Task finished name='Task-32' coro=<AsyncServer._handle_event_internal() done, defined at /Users/max/Library/Caches/pypoetry/virtualenvs/reflex-D4pZYT2f-py3.12/lib/python3.12/site-packages/socketio/async_server.py:605> exception=AttributeError("'NoneType' object has no attribute 'router'")>
Traceback (most recent call last):
  File "/Users/max/Library/Caches/pypoetry/virtualenvs/reflex-D4pZYT2f-py3.12/lib/python3.12/site-packages/socketio/async_server.py", line 607, in _handle_event_internal
    r = await server._trigger_event(data[0], namespace, sid, *data[1:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/max/Library/Caches/pypoetry/virtualenvs/reflex-D4pZYT2f-py3.12/lib/python3.12/site-packages/socketio/async_server.py", line 643, in _trigger_event
    return await handler.trigger_event(event, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/max/Library/Caches/pypoetry/virtualenvs/reflex-D4pZYT2f-py3.12/lib/python3.12/site-packages/socketio/async_namespace.py", line 37, in trigger_event
    ret = await handler(*args)
          ^^^^^^^^^^^^^^^^^^^^
  File "/Users/max/projects/senbax/reflex-maxim/reflex/app.py", line 1514, in on_event
    async for update in process(self.app, event, sid, headers, client_ip):
  File "/Users/max/projects/senbax/reflex-maxim/reflex/app.py", line 1273, in process
    state.router = RouterData(router_data)
    ^^^^^^^^^^^^
  File "/Users/max/projects/senbax/reflex-maxim/reflex/state.py", line 1099, in __setattr__
    setattr(self.parent_state, name, value)
AttributeError: 'NoneType' object has no attribute 'router'

```

masenf replied:
that error arises because the event is triggered without the parent state prefix

* `"frontend_event_exception_state.handle_frontend_exception"` :x:
* `"state.frontend_event_exception_state.handle_frontend_exception"` ✅

maximvlah replied:
thanks! I just fixed it

maximvlah commented:
I was able to fix the issue mentioned by @masenf with error boundary wrapper.

I used the **react-error-boundary** lib recommended by React team: 
https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary

This allows us to catch unhandled errors and render a fallback component. To test it out, I have added **useErrorBoundary** variable that we can pass to any component if we want to wrap it and its children with the error boundary, so in your case it would be:

```python
def index() -> rx.Component:
    return rx.container(
        rx.vstack(
            rx.cond(
                State.broken,
                rx.text("Broken", size=rx.Var.create("x", _var_is_local=False, _var_is_string=False)),
            ),
            rx.text("Break the app? ", rx.switch(checked=State.broken, on_change=State.set_broken)),
            spacing="5",
            justify="center",
            min_height="85vh",
            useErrorBoundary=True, # Wraps rx.vstack and all of its children with the default ErrorBoundary
        ),
        rx.logo(),
    )
```

This made it catch these errors in production too, however in debug mode it fired the error handler thrice. So, I made the error boundary call onError only in production through checking the `process.env.NODE_ENV`. Still we have the bug with frontend errors fired twice in debug mode which according to https://github.com/facebook/react/issues/11499 is a known issue with no trivial solution. 

Still the ErrorBoundary also has limitations, i.e. does not catch all unhandled errors. https://legacy.reactjs.org/docs/error-boundaries.html

Therefore, in my opinion, keeping the `window.onerror` logic which I introduced initially may enable us to catch most of the frontend errors. This is all hacky and I am not fully satisfied with this solution, however I haven’t found a better one yet. If we go for it, it would be great to allow users to define the fallback component as well as `onError` and `onReset` behaviour from python side. I followed the docs and tried wrapping ErrorBoundary in Reflex, however had little success, especially with the callbacks. I would be very grateful if you could help me properly wrap it. This is what I tried:

```python
class ErrorBoundary(Component):
	
	library = "react-error-boundary"
	tag = "ErrorBoundary"
	
	# Fallback component must take error and resetErrorBoundary props
	# https://github.com/bvaughn/react-error-boundary?tab=readme-ov-file#errorboundary-with-fallbackcomponent-prop
	fallback: rx.Var[Component] = rx.text("Unknown Error")
	# https://github.com/bvaughn/react-error-boundary?tab=readme-ov-file#logging-errors-with-onerror
	on_error: rx.EventHandler
	on_reset: rx.EventHandler
```

Regarding the errors mentioned by @Lendemor, I think they both occur before the event loop (on which all my logic depends) is instantiated and thus the frontend error events are not propagated to the python side.

Please, check out the latest commit I made just now and let me know what you think.

maximvlah commented:
@benedikt-bartscher I must have missed it, just addded backend exception handler to _process, thanks for pointing it out. 
I think it is better to keep telemetry separate as if the user overrides the `backend exception handler` and the telemetry is enabled, then it would not be sent to reflex team.

benedikt-bartscher commented:
https://github.com/reflex-dev/reflex/pull/3476/commits/6cd628f987b0733d9b5b506dcb1c46aee5ce1d80 introduced a lot of formatting, iirc this has already been done on main. Please rebase/merge main or cleanup this commit

Lendemor commented:
> [6cd628f](https://github.com/reflex-dev/reflex/commit/6cd628f987b0733d9b5b506dcb1c46aee5ce1d80) introduced a lot of formatting, iirc this has already been done on main. Please rebase/merge main or cleanup this commit

This changes are not on main, but they shouldn't be because we've added that rule to `lint.ignore` in `pyproject.toml`.
Not sure how `ruff` was run on that PR, but it should pick up `pyproject.toml` normally.

maximvlah commented:
> Please rebase/merge main or cleanup this commit

@benedikt-bartscher Done

Lendemor commented:
There is still too many modified files, the additional blank lines in the docs shouldn't be there.

maximvlah commented:
@Lendemor @benedikt-bartscher ok, it looks like it is clean now, sorry for the mess. I want to rename the branch to a more meaningful name. I have just tried it from github web  like [here](https://github.blog/changelog/2021-01-19-support-for-renaming-an-existing-branch/) however it closed the PR. Is there another way or should I just leave like this?

masenf commented:
@maximvlah  i'm still seeing the blank line changes in many files. you might try just opening a new PR with your renamed and fixed branch and i'll test it out and review it. we really want to get this feature in and appreciate you working on it 😄

maximvlah commented:
@masenf ok, will do it now
